Add LevelDB provider.

eslachance opened the pull request:
### Proposed Semver Increment Bump: [MAJOR/MINOR/PATCH]
minor

### Changes Proposed in this Pull Request (List new items in CHANGELOG.MD)

- Added `level` provider
- 2 new dependencies: level and then-level (promisified)

eslachance committed:
Merge pull request #1 from dirigeants/indev

Indev

eslachance committed:
Merge pull request #2 from dirigeants/indev

Indev

eslachance committed:
Leveldb provider

Support for LevelDB provider, which should be a drop-in replacement for json. Currently NOT compatible with standard sharding.

kyranet commented on providers/level.js:
```diff
@@ -0,0 +1,146 @@
+};
+
+exports.conf = {
+  moduleName: "json",
```
Nice copypaste.

eslachance replied:
fixed

kyranet commented on providers/level.js:
```diff
@@ -0,0 +1,146 @@
+
+exports.shutdown = async () => {
+  await loaded.forEach(async db => db.close());
+};
```
You could remove that async/await and make the function return loaded.forEach()

```js
exports.shutdown = () => loaded.forEach(...);
```

eslachance replied:
Wouldn't that mean the shutdown would resolve before all dbs are closed?

kyranet replied:
Actually, not. Because returning a forEach (sync method) will freeze the event loop until it finishes, and it finishes when all the iterations have been executed. As well as you don't need `async db => db.close()`, if db.close returns a promise, it the function returns a promise. (What async does in there, but slowing down the execution)

eslachance replied:
fixed

kyranet commented on providers/level.js:
```diff
@@ -0,0 +1,146 @@
+ * @param {string} table The name of the directory.
+ * @returns {Promise<Object>}
+ */
+exports.getRandom = () => null;
```
How is this not implementable? Doesn't getAll return an array of objects?

eslachance replied:
We're not caching results which means I would need to re-load every single result, grab a random, and let node garbage collect all these results. isn't that super bad in terms of performance? why do we need random anyway?

kyranet replied:
Then, remove it, lol.

eslachance replied:
done

eslachance committed:
Changes pertainin to Kira's comments

removed random, simplified shutdown. Also, moved .create() aliases to a better suited location. And ensuredir *before* trying to read that dir...

kyranet commented on providers/level.js:
```diff
@@ -0,0 +1,140 @@
+
+/**
+ * Closes all open database connections.
+ * @returns {Array<Promise>}
```
Any forEach will return nothing.

```js
* @returns {Promise<Void>}
```

eslachance replied:
fixed

kyranet commented on providers/level.js:
```diff
@@ -0,0 +1,140 @@
+ * @returns {Promise<Object[]>}
+ */
+exports.getAll = async (table) => {
+  if (!loaded.has(table)) return {};
```
Should return an empty array, not an object.

eslachance replied:
Fixed, also fixed the inserting that wasn't previously adding the ID to it. Or stringifying.

kyranet commented on providers/level.js:
```diff
@@ -0,0 +1,140 @@
+exports.createTable = async (table) => {
+  const db = await promisify(level(baseDir + sep + table));
+  loaded.set(table, db);
+  return null;
```
This return is not needed. And you could do:

```js
exports.createTable = table => promisify(level(baseDir + sep + table))
  .then(db => loaded.set(table, db));
```

eslachance replied:
Done

kyranet commented on providers/level.js:
```diff
@@ -0,0 +1,140 @@
+exports.conf = {
+  moduleName: "level",
+  enabled: true,
+  requiredModules: [],
```
You missed the required modules to run this provider. `fs-nextra`, `level` and `then-levelup`

eslachance replied:
fixed

kyranet reviewed (changes requested):
More stuff I found on my PC.

UnseenFaith committed:
Merge branch 'indev' into indev

eslachance committed:
fixed per Kyra's comments

all done

eslachance committed:
Merge remote-tracking branch 'origin/indev' into indev

eslachance commented:
Closed because this belongs in komada-pieces
